Draft doc on using seqset API (#870)
pathoplexus/pathoplexus: monorepo/website/src/pages/docs/how-to/generate-seqset-api.mdx

---
layout: ../../../layouts/DocLayout.astro
title: "Generating SeqSets via the API"
order: 51
---
import InfoBox from '../../../components/asides/InfoBox.astro';

### Contents
- [Creating SeqSets](#Creating-seqsets)
- [Updating SeqSets](#Updating-seqsets)
- [Create SeqSet doi](#Create-seqset-doi)

In order to use the Pathoplexus API for creating SeqSets, please ensure that you have read the [Data Use Terms](/about/terms-of-use/data-use-terms)
and agree to them. For detailed information on how to use and interact with the APIs please checkout the <em>swagger UI</em>
documentation. Here we shall demonstrate how to use curl from the terminal to generate SeqSets using the
`POST`, `PUT`, and `GET` http methods. To effectively interact with the API, you will need to learn about
these methods as well as the endpoints.

> A full list of endpoints for SeqSets can be found under the `seqset-citations-controller` section at 

Reviewer: ```suggestion
A full list of endpoints for SeqSets can be found under the `seqset-citations-controller` section in our
```

[Swagger UI](https://backend.pathoplexus.org/swagger-ui/index.html) but here we shall focus on 
the `/create-seqset`, `/update-seqset`, `create-seqset-doi`. 

<InfoBox>
**Important**
Please note that the SeqSet endpoints require authentication or a bearer token. The details for authenticating
via API or retrieving a token from the browser are explained [here](https://pathoplexus.org/docs/how-to/authentication-api)
</InfoBox>

## Creating seqsets
A SeqSet can contain both focal and background sequences [explanation of the differences here](https://pathoplexus.org/about/terms-of-use/data-use-terms).
To define this in our request we shall need to set the <em>isFocal</em> parameter to true or false on our data object in 

> the request body. The other parametes that we need to provide include the accession numbers of the sequences

Reviewer: ```suggestion
the request body. The other parameters that we need to provide include the accession numbers of the sequences
```

> and the type. Please checkout the schema on the Swagger UI to understand the accepted data formats. Below is

Reviewer: ```suggestion
and the type. Please check the schema on the Swagger UI to understand the accepted data formats. Below is
```

example code for creating a SeqSet for 2 sequences.

```
curl -X 'POST' 'https//backend.pathoplexus.org/create-seqset' \
     -H 'accept: application/json' \
     -H 'Authorization: Bearer <TOKEN> \
     -H 'Content-Type: application/json \
     -d '{
        "name": "Add name for your seqset",
        "description": "Add a description for your seqset",
        "records": [
            {
                "accession": "PP_12345",
                "type": "Loculus",
                "isFocal": true

            },
            {
                "accession": "PP_654321",
                "type": "Loculus",
                "isFocal": false
            }
        ]
     }
```

If the SeqSet is was created successully you will get a 200 OK response code with the SeqSet ID and version number

> for the created seqset. Otherwise, an error code will be returned, for a description of the different error codes see the 

Reviewer: ```suggestion
for the created seqset. Otherwise, an error code will be returned; for a description of the different error codes see the 
```

[Swagger UI](https://backend.pathoplexus.org/swagger-ui/index.html#/seq-set-citations-controller/createSeqSet).

<br>
**Note**

> Please remember to replace your `TOKEN` in the header.

Reviewer: ```suggestion
Please remember to replace your `TOKEN` in the header, and to replace the `name` and `description` with the information about your own SeqSet.
```

</br>

## Updating seqsets

> Once you have created a SeqSet, you are able to update it, for example by adding additional sequences. This will not modify the SeqSet but rather it

Reviewer: ```suggestion
Once you have created a SeqSet, you are able to update it, for example by adding additional sequences. This 
```
I think it's otherwise a bit confusing to say 'it will not modify the SeqSet' - as people might think it's not changing anything. It's true it doesn't change the older version, but we should try to make this very clear to people.

Author: I take note of this and understand the reason for the change

> will create a new version of that SeqSet, with the same Seqset ID. Below is an example for updating SeqSets.

Reviewer: ```suggestion
will create a new version of that SeqSet, with the same Seqset ID. The older version will show the original set of sequences included. Below is an example for updating SeqSets.
```

> Take note of the additional parameters in the request body to include the `seqSetID`.

Reviewer: Should one give the same SeqSet name and description? If you include new values does it update? or error? We should be clear about this. Can you leave out the name and description if you don't want to change these values?

```
curl -X 'PUT' 'https//backend.pathoplexus.org/create-seqset' \
     -H 'accept: application/json' \
     -H 'Authorization: Bearer <TOKEN> \
     -H 'Content-Type: application/json \
     -d '{
        "seqSetId": "PP_SS_XX"
        "name": "Add name for your seqset",
        "description": "Add a description for your seqset",
        "records": [
            {
                "accession": "PP_12345",
                "type": "Loculus",
                "isFocal": true

            },
            {
                "accession": "PP_654321",
                "type": "Loculus",
                "isFocal": false
            },
            {
                "accession": "PP_334422",
                "type": "Loculus",
                "isFocal": true
            }
        ]
     }
```
If the SeqSet is updated successfully, you will get the 200 OK response with the SeqSet ID and the new version number.

## Create SeqSet DOI

> After you have a created a SeqSet you can be able to generate a DOI for it. In order to this, you will need to

Reviewer: ```suggestion
After you have a created a SeqSet you can be able to generate a DOI for it. (As required by the [Data Use Terms when using Restricted Use data in some circumstances](https://pathoplexus.org/about/terms-of-use/data-use-terms#422-publications-and-preprints).) In order to this, you will need to
```

parse the SeqSet ID and version of the seqset as parameters to the backend URL using the `POST` method. 

```
curl -X 'POST' 'https//backend.pathoplexus.org/create-seqset-doi?seqSetID=PP_SS_XX&version=1' \
     -H 'accept: application/json' \
     -H 'Authorization: Bearer <TOKEN> \
     -d ''
```
<br>
**Important**

> Please note that after a DOI has been created for a SeqSet, it cannot nolonger be deleted.

Reviewer: ```suggestion
Please note that after a DOI has been created for a SeqSet, it cannot no longer be deleted.
```

     -d '{
        "name": "Add name for your seqset",
        "description": "Add a description for your seqset",
        "records": [
            {
                "accession": "PP_12345",
                "type": "Loculus",
                "isFocal": true

            },
            {
                "accession": "PP_654321",
                "type": "Loculus",
                "isFocal": false
            }
        ]
     }
```

If the seqset is was created successully you will get a 200 OK response code with the seqset ID and version number
for the created seqset. Otherwise, a range of error codes maybe returned and a full description can be found at 
[Swagger UI](https://backend.pathoplexus.org/swagger-ui/index.html).

<br>
**Note**
Please remember to replace `x-request-id` and your `TOKEN` in the header.
</br>

## Updating seqsets
Once you have created a seqset, you are able to update it with additional sequences that you could have added later
after the seqset was already created. You need to know that this wont replace the previous seqset but rather it
will create a new one and with version number 2, but maintain the Seqset ID. Below is an example for updating seqsets.

> Take note of the additional parameters in the request body to include the `seqSetID`.

Reviewer 2: ```suggestion
Once you have created a SeqSet, you are able to update it, for example by adding additional sequences. This will not modify the SeqSet but rather it
will create a new version of that SeqSet, with the same Seqset ID. Below is an example for updating SeqSets.
Take note of the additional parameters in the request body to include the `seqSetID`.
```

     -H 'Authorization: Bearer <TOKEN> \
     -H 'Content-Type: application/json \
     -d '{
        "seqSetId": "PP_SS_XX"
        "name": "Add name for your seqset",
        "description": "Add a description for your seqset",
        "records": [
            {
                "accession": "PP_12345",
                "type": "Loculus",
                "isFocal": true

            },
            {
                "accession": "PP_654321",
                "type": "Loculus",
                "isFocal": false
            },
            {
                "accession": "PP_334422",
                "type": "Loculus",
                "isFocal": true
            }
        ]
     }
```
If the seqset is updated successfully, you will get the 200 OK response with the seqsetID and the new version number.

## Create seqset doi
After you have a created a seqset you can be able to generate a doi for it. In order to this, you will need to
parse the seqsetID and version of the seqset as parameters to the backend URL using the `POST` method. 

```
curl -X 'POST' 'https//backend.pathoplexus.org/create-seqset-doi?seqSetID=PP_SS_XX&version=1' \
     -H 'accept: application/json' \

>      -H 'x-request-id: <replace with your UUID> \

Reviewer 2: ```suggestion
```

- [Validating SeqSet records](#Validating-SeqSet-records)

In order to use the Pathoplexus API for creating SeqSets, please ensure that you have read the [Data Use Terms](/about/terms-of-use/data-use-terms)
and agree to them. For detailed information on how to use and interact with the APIs please checkout the <em>swagger UI</em>
documentation. Here we shall demonstrate how to use curl from the terminal to generate SeqSets using the
`POST`, `PUT`, `DELETE`, and `GET` http methods. To effectively interact with the API, you will need to learn about
these methods as well as the endpoints.

A full list of endpoints for SeqSets can be found under the `seqset-citations-controller` section in our
[Swagger UI](https://backend.pathoplexus.org/swagger-ui/index.html) but here we shall focus on 
the `/create-seqset`, `/update-seqset`, `create-seqset-doi`. 

<InfoBox>
<b>Important</b>
<p>Please note that the SeqSet endpoints require authentication or a bearer token. The details for authenticating via API or retrieving a token from the browser are explained <a href="https://pathoplexus.org/docs/how-to/authentication-api">here</a></p>
</InfoBox>

> ## Creating seqsets

Reviewer 2: ```suggestion
## Creating SeqSets
```

To define this in our request we shall need to set the <em>isFocal</em> parameter to true or false on our data object in 
the request body. The other parameters that we need to provide include the accession numbers of the sequences
and the type. Please check the schema on the Swagger UI to understand the accepted data formats. Below is
example code for creating a SeqSet for 2 sequences.

```
curl -X 'POST' 'https//backend.pathoplexus.org/create-seqset' \
     -H 'accept: application/json' \
     -H 'Authorization: Bearer <TOKEN> \
     -H 'Content-Type: application/json \
     -d '{
        "name": "Add name for your seqset",
        "description": "Add a description for your seqset",
        "records": [
            {
                "accession": "PP_12345",
                "type": "Loculus",
                "isFocal": true

            },
            {
                "accession": "PP_654321",
                "type": "Loculus",
                "isFocal": false
            }
        ]
     }
```
If the SeqSet is was created successully you will get a 200 OK response code with the SeqSet ID and version number

> for the created seqset. Otherwise, an error code will be returned; for a description of the different error codes see the 

Reviewer 2: ```suggestion
for the created SeqSet. Otherwise, an error code will be returned; for a description of the different error codes see the 
```

     -H 'accept: application/json' \
     -H 'Authorization: Bearer <TOKEN> \
     -H 'Content-Type: application/json \
     -d '{
        "seqSetId": "PP_SS_XX"
        "name": "Add name for your seqset",
        "description": "Add a description for your seqset",
        "records": [
            {
                "accession": "PP_12345",
                "type": "Loculus",
                "isFocal": true

            },
            {
                "accession": "PP_654321",
                "type": "Loculus",
                "isFocal": false
            },
            {
                "accession": "PP_334422",
                "type": "Loculus",
                "isFocal": true
            }
        ]
     }
```
If the SeqSet is updated successfully, you will get the 200 OK response with the SeqSet ID and the new version number.

## Create SeqSet DOI
After you have a created a SeqSet you can be able to generate a DOI for it. (As required by the [Data Use Terms when using Restricted Use data in some circumstances](https://pathoplexus.org/about/terms-of-use/data-use-terms#422-publications-and-preprints).) In order to this, you will need to

> parse the SeqSet ID and version of the seqset as parameters to the backend URL using the `POST` method. 

Reviewer 2: ```suggestion
pass the SeqSet ID and version of the SeqSet as parameters to the backend URL using the `POST` method. 
```

```
curl -X 'POST' 'https//backend.pathoplexus.org/create-seqset-doi?seqSetID=PP_SS_XX&version=1' \
     -H 'accept: application/json' \
     -H 'Authorization: Bearer <TOKEN> \
     -d ''
```
Please note that after a DOI has been created for a SeqSet, it cannot no longer be deleted.

> Therefore, you may want to create a `DOI` after all the necessary updates or changes to your SeqSet have been made

Reviewer 2: ```suggestion
Therefore, you may want to create a `DOI` after all the necessary updates or changes to your SeqSet have 
```

> been made. After, the older versions of your `SeqSet` can then be deleted using the `DELETE` method as shown in the example below. 

Reviewer 2: maybe we can make deletion a separate section before creating DOIs
